### PR TITLE
[youtube] Redefine uploader metadata

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

`uploader` metadata for videos, tabs, etc. is/was inconsistent and not well defined. With handles and the channel UCID url now being the two primary urls to a channel, this updates the metadata to reflect that.

New mapping will be strictly:
```
channel -> channel name
channel_id -> Only UCID
channel_url ->Only  UCID /channel url

uploader -> channel name (same as channel field)
uploader_id -> Only @handle
uploader_url - > Only @handle url 
```

Note this is slightly different than https://github.com/ytdl-org/youtube-dl/commit/f7ce98a21e15cb094c772e9082796d009c61578b

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by pukkandan** [comment]
cc @dirkf

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -451,6 +451,23 @@ class YoutubeBaseInfoExtractor(InfoExtractor):
 
     _IGNORED_WARNINGS = {'Unavailable videos will be hidden during playback'}
 
+    _YT_HANDLE_RE = r'@[A-Za-z0-9_.-]{3,30}'  # https://support.google.com/youtube/answer/11585688?hl=en
+    _YT_CHANNEL_UCID_RE = r'UC[A-Za-z0-9_-]{22}'
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:455]:
```suggestion
    _YT_HANDLE_RE = r'@[\w.-]{3,30}'  # https://support.google.com/youtube/answer/11585688?hl=en
    _YT_CHANNEL_UCID_RE = r'UC[\w-]{22}'
```

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -451,6 +451,23 @@ class YoutubeBaseInfoExtractor(InfoExtractor):
...
+
+    @classmethod
+    def ucid_or_none(cls, ucid):
+        return ucid if re.match(rf'^{cls._YT_CHANNEL_UCID_RE}$', ucid or '') is not None else None
+
+    @classmethod
+    def handle_or_none(cls, handle):
+        return handle if re.match(rf'^{cls._YT_HANDLE_RE}$', handle or '') is not None else None
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:463]:
```suggestion

    @classmethod
    def ucid_or_none(cls, ucid):
        return self._search_regex(rf'^{cls._YT_CHANNEL_UCID_RE}$', ucid, 'UC-id', default=None)

    @classmethod
    def handle_or_none(cls, handle):
        return self._search_regex(rf'^{cls._YT_HANDLE_RE}$', handle, '@-handle', default=None)
```

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -451,6 +451,23 @@ class YoutubeBaseInfoExtractor(InfoExtractor):
...
+
+    def handle_from_url(self, url):
+        return self._search_regex(rf'https?://(?:www\.)?youtube\.com/({self._YT_HANDLE_RE})', urljoin('https://www.youtube.com/', url) or '', 'channel handle', default=None)
+
+    def ucid_from_url(self, url):
+        return self._search_regex(rf'https?://(?:www\.)?youtube\.com/channel/({self._YT_CHANNEL_UCID_RE})', urljoin('https://www.youtube.com/', url) or '', 'channel id', default=None)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:469]:
```suggestion
    def handle_from_url(self, url):
        return self._search_regex(rf'(?:https?://(?:www\.)?youtube\.com/)?({self._YT_HANDLE_RE})',
                                  url, 'channel handle', default=None)

    def ucid_from_url(self, url):
        return self._search_regex(rf'(?:https?://(?:www\.)?youtube\.com/)?({self._YT_CHANNEL_UCID_RE})',
                                  url, 'channel id', default=None)
```

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -4118,6 +4072,11 @@ def is_bad_format(fmt):
 
         formats.extend(self._extract_storyboard(player_responses, duration))
 
+        channel_handle = self.handle_from_url(owner_profile_url)
+        if not channel_handle:
+            # fallbacks
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4077]: TODO

File: yt_dlp/extractor/youtube.py [hunk 5]
```diff
@@ -4482,18 +4432,28 @@ def _extract_basic_item_renderer(item):
...
         title = self._get_text(renderer, 'title')
-        channel_url = f'https://www.youtube.com/channel/{channel_id}'
+        channel_url = format_field(channel_id, None, 'https://www.youtube.com/channel/%s', default=None)
+        # As of 2023-03-01 YouTube doesn't use the channel handles on these renderers yet.
+        # However we can expect them to change that in the future.
+        channel_handle = self.handle_from_url(
+            traverse_obj(renderer, (
+                'navigationEndpoint',
+                (('commandMetadata', 'webCommandMetadata', 'url'), ('browseEndpoint', 'canonicalBaseUrl'))),
+                         get_all=False,expected_type=str))
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4444]:
```suggestion
            traverse_obj(renderer, (
                'navigationEndpoint', (('commandMetadata', 'webCommandMetadata', 'url'),
                                       ('browseEndpoint', 'canonicalBaseUrl')),
                {str}), get_all=False))
```

File: yt_dlp/extractor/youtube.py [hunk 6]
```diff
@@ -4833,13 +4793,13 @@ def _extract_metadata_from_tabs(self, item_id, data):
 
         metadata_renderer = traverse_obj(data, ('metadata', 'channelMetadataRenderer'), expected_type=dict)
         if metadata_renderer:
+            channel_id = self.ucid_or_none(metadata_renderer.get('externalId')) or self.ucid_from_url(metadata_renderer.get('channelUrl'))
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4796]:
```suggestion
            channel_id = traverse_obj(metadata_renderer, ('externalId', {self.ucid_or_none}),
                                                         ('channelUrl', {self.ucid_from_url}))
```

File: yt_dlp/extractor/youtube.py [hunk 7]
```diff
@@ -4892,6 +4852,13 @@ def _get_uncropped(url):
             'thumbnails': (primary_thumbnails or playlist_thumbnails) + avatar_thumbnails + channel_banners,
         })
 
+        channel_handle = (self.handle_from_url(traverse_obj(metadata_renderer, 'vanityChannelUrl')  # also try ownerUrls from metadata renderer (channel)
+                          or self.handle_or_none(traverse_obj(data, ('header', ..., 'channelHandleText')))))
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4856]:
```suggestion
        channel_handle = (self.handle_from_url(traverse_obj(metadata_renderer, 'vanityChannelUrl'))  # TODO: also try ownerUrls from metadata renderer (channel)
                          or traverse_obj(data, ('header', ..., 'channelHandleText', {self.handle_or_none}), get_all=False))
```

File: yt_dlp/extractor/youtube.py [hunk 8]
```diff
@@ -4923,16 +4890,17 @@ def _get_uncropped(url):
...
+                'uploader_id': self.handle_from_url(urljoin('https://www.youtube.com', browse_ep.get('canonicalBaseUrl')))
             })
 
         info.update({
-            'channel': info['uploader'],
-            'channel_id': info['uploader_id'],
-            'channel_url': info['uploader_url']
+            'uploader': info['channel'],
+            'channel_url': format_field(info.get('channel_id'), None, 'https://www.youtube.com/channel/%s', default=None),
+            'uploader_url': format_field(info.get('uploader_id'), None, 'https://www.youtube.com/%s',default=None),
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:4901]:
```suggestion
            'uploader_url': format_field(info.get('uploader_id'), None, 'https://www.youtube.com/%s', default=None),
```

**Comment by dirkf** [comment]
I compared the extractor test-cases. Where yt-dlp specifies uploader metadata here, the yt-dl test-cases in the latest git master generally match if present. Here are the non-matching ones.

YoutubeIE:
```diff
--- yt-dl
+++ yt-dlp
 ...
             'url': 'OtqTfy26tG0',
 ...
-                'uploader': 'The Cinematic Orchestra - Topic',
+                'uploader': 'The Cinematic Orchestra',
 ...
```
I didn't change this result, which is Sergey's from 0cf09c2b41, though the yt-dlp result looks more plausible.

I think these remaining differences are all results that I changed in f7ce98a21e15cb094c772e9082796d009c61578b.

YoutubePlaylistIE:
```diff
--- yt-dl
+++ yt-dlp
 ...
        'url': 'PLBB231211A4F62143',
 ...
-                'uploader_id': '@WickmanVT',
+                'uploader_id': 'UCKSpbfbl5kRQpTdL7kMc-1Q',
 ...
         'url': 'https://www.youtube.com/embed/videoseries?list=PL6IaIsEjSbf96XFRuNccS_RuEXwNdsoEu',
 ...
             'id': 'PL6IaIsEjSbf96XFRuNccS_RuEXwNdsoEu',
 ...
-                'uploader_id': '@milan5503',
+                'uploader_id': 'UCEI1-PVPcYXjB73Hfelbmaw',
 ...
         'url': 'http://www.youtube.com/embed/_xDOZElKyNU?list=PLsyOSbh5bs16vubvKePAQ1x3PhKavfBIl',
 ...
-                'uploader_id': '@music_king',
+                'uploader_id': 'UC21nz3_MesPLqtDqwdvnoxA',
 ...
```
YoutubeYtBeIE:
```diff
--- yt-dl
+++ yt-dlp
 ...
         'url': 'https://youtu.be/yeWKywCrFtk?list=PL2qgrgXsNUG5ig9cat4ohreBjYLAPC0J5',
 ...
-                'uploader_id': '@backuspagemuseum',
+                'uploader_id': 'backuspagemuseum',
 ...
```

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (dirkf) [reply yt_dlp/extractor/youtube.py:455]: Is YT actually sending strings with Unicode alphanumerics outside `[a-zA-Z\d]`?

Comment (coletdjnz) [reply yt_dlp/extractor/youtube.py:455]: > Is YT actually sending strings with Unicode alphanumerics outside `[a-zA-Z\d]`?

https://www.youtube.com/channel/UCiu-3thuViMebBjw_5nWYrA

https://wiki.archiveteam.org/index.php/YouTube/Technical_details

**Comment by coletdjnz** [comment]
TODO:
- investigate channel/uploader name extraction (might be getting wrong one for topic channels)
- handle extraction fallback for channel pages (ownerUrls)
- notificationsIE

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (dirkf) [reply yt_dlp/extractor/youtube.py]: Sure, I wasn't thinking of the punctuation-like characters. So either `(?A)[\w.-]` (loses in Py2) or just unchanged. The ArchiveTeam link is excellent.

**Comment by dirkf** [comment]
Reviewing these tests for a new yt-dl extractor version, I find a contradiction between the `channel` ("Full name of the channel the video is uploaded on.") values extracted for test videos IB3lcPjvWLA and MgNrAu2pzNs (eg).

IB3lcPjvWLA: `Afrojack` (yt-dlp test value) vs `AfrojackVEVO`
MgNrAu2pzNs: `Stephen` (yt-dlp test value) vs `Stephen - Topic`

Currently the yt-dl extractor doesn't test these values because it just [sets `channel` from `uploader`, if set](https://github.com/ytdl-org/youtube-dl/issues/31530#issuecomment-1435734719).

The second test is commented `TODO: should be "Stephen - Topic"`. If so, I believe that the corresponding value `AfrojackVEVO` must be correct for IB3lcPjvWLA, which would (FWIW) match yt-dl's current values and the header of the corresponding channel page in Invidious in [each](https://vid.puffyan.us/channel/UChuZAo1RKL85gev3Eal9_zg) [case](https://vid.puffyan.us/channel/UC-pWHpBjdGG69N9mM2auIAA).

[Separately, what are these ` - Topic` suffixes?]

The first value is here (similar to where yt-dlp gets it)
```py
traverse_obj(initial_data, (
    'contents', 'twoColumnWatchNextResults', 'results', 'results',
    'contents', ..., lambda k, _: k.endswith('SecondaryInfoRenderer'),
    ('owner', 'videoOwner'), 'videoOwnerRenderer', 'title'),
    get_all=False)
```
and here
```py
traverse_obj(initial_data, (
    'engagementPanels', ..., 'engagementPanelSectionListRenderer',
    'content', 'structuredDescriptionContentRenderer', 'items', ...,
    'videoDescriptionHeaderRenderer', 'channel'),
    get_all=False)
```
Everywhere else, eg `traverse_obj(microformats, (..., 'ownerChannelName'))` and as in the linked post, matches the second value.

What should be the correct resolution of this `TODO`?

I expect a similar issue applies to tab extraction of video data for playlists. The `channel` value extracted there should be consistent.

**Comment by coletdjnz** [comment]
An issue should be raised for this.

The topic suffix is correct unless YouTube has changed the channel. It's a topic channel which are quite messy to deal with.

**Pull request merged**